### PR TITLE
[actpool] persistent storage for actions

### PR DESCRIPTION
**Pull request by envestcc**
# Description

Now actions in actpool will be persistent on disk. There are three main functionalities:
- persistence: It will save an action when it's accepted by actpool, with configureable datacap
- load-on-start: all persistent actions will be load at actpool start
- reload periodically: actions dropped in memory will be attemped to reload into memory periodically

Based on #4362 

## Type of change
Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: chainservice/builder.go [hunk 1]
```diff
@@ -245,7 +246,29 @@ func (builder *Builder) createElectionCommittee() (committee.Committee, error) {
...
+		if builder.cfg.ActPool.Store != nil {
+			evmID := builder.cfg.Chain.EVMNetworkID
+			options = append(options, actpool.WithStore(
+				*builder.cfg.ActPool.Store,
+				func(selp *action.SealedEnvelope) ([]byte, error) {
+					return proto.Marshal(selp.Proto())
+				},
+				func(blob []byte) (*action.SealedEnvelope, error) {
+					d := &action.Deserializer{}
+					d.SetEvmNetworkID(evmID)
```
Comment (CoderZhi) [inline chainservice/builder.go:259]: move out to line 252

File: actpool/blobstore.go [hunk 2]
```diff
@@ -0,0 +1,245 @@
+package actpool
```
Comment (CoderZhi) [inline actpool/blobstore.go:1]: this is not a blobstore, but a general store

**Review by CoderZhi — COMMENTED** [review]
the others look good to me

**Comment by dustinxie** [comment]
as discussed, a general comment is we should only enable this for blob tx, b/c each tx is 128kB and each block can take only 2 blobs, leading to a potential backlog of such txs.
For regular tx (size is ~100 bytes, and don't have limitation in each block), it is not needed to persist them locally. Because after a restart, the stored txs are all either already added to past blocks, or purged due to processing error (like wrong chainID), seems useless to me to persist these regular tx

File: actpool/actionstore.go [hunk 3]
```diff
@@ -68,24 +68,24 @@ var (
 	errStoreNotOpen = fmt.Errorf("blob store is not open")
 )
 
-var defaultBlobStoreConfig = blobStoreConfig{
-	Datadir: "blobpool",
+var defaultActionStoreConfig = actionStoreConfig{
+	Datadir: "actionstore",
 	Datacap: 10 * 1024 * 1024 * 1024,
```
Comment (dustinxie) [inline actpool/actionstore.go:73]: is this 10GB?

File: actpool/actionstore.go [hunk 4]
```diff
@@ -68,24 +68,24 @@ var (
...
 }
 
-func newBlobStore(cfg blobStoreConfig, encode encodeAction, decode decodeAction) (*blobStore, error) {
+func newBlobStore(cfg actionStoreConfig, encode encodeAction, decode decodeAction) (*actionStore, error) {
```
Comment (dustinxie) [inline actpool/actionstore.go:76]: newActionStore

File: actpool/actpool.go [hunk 5]
```diff
@@ -345,6 +381,16 @@ func (ap *actPool) GetUnconfirmedActs(addrStr string) []*action.SealedEnvelope {
 func (ap *actPool) GetActionByHash(hash hash.Hash256) (*action.SealedEnvelope, error) {
 	act, ok := ap.allActions.Get(hash)
 	if !ok {
+		if ap.store != nil {
```
Comment (dustinxie) [inline actpool/actpool.go:384]:
```
if ok {
    return act.(*action.SealedEnvelope), nil
}
if ap.store != nil {
}
```

File: actpool/actpool.go [hunk 6]
```diff
@@ -412,6 +458,21 @@ func (ap *actPool) validate(ctx context.Context, selp *action.SealedEnvelope) er
 	return nil
 }
 
+func (ap *actPool) removeReplacedActs(acts []*action.SealedEnvelope) {
```
Comment (dustinxie) [inline actpool/actpool.go:461]: this is same as `removeInvalidActs`, see comment below?

File: actpool/actpool.go [hunk 7]
```diff
@@ -412,6 +458,21 @@ func (ap *actPool) validate(ctx context.Context, selp *action.SealedEnvelope) er
...
+		}
+		log.L().Debug("Removed replaced action.", log.Hex("hash", hash[:]))
+		ap.allActions.Delete(hash)
+		intrinsicGas, _ := act.IntrinsicGas()
+		atomic.AddUint64(&ap.gasInPool, ^uint64(intrinsicGas-1))
+		ap.accountDesActs.delete(act)
+	}
+}
+
 func (ap *actPool) removeInvalidActs(acts []*action.SealedEnvelope) {
```
Comment (dustinxie) [inline actpool/actpool.go:476]: add a bool input `removeInvalid`

File: actpool/actpool.go [hunk 8]
```diff
@@ -424,6 +485,11 @@ func (ap *actPool) removeInvalidActs(acts []*action.SealedEnvelope) {
 		intrinsicGas, _ := act.IntrinsicGas()
 		atomic.AddUint64(&ap.gasInPool, ^uint64(intrinsicGas-1))
 		ap.accountDesActs.delete(act)
+		if ap.store != nil {
```
Comment (dustinxie) [inline actpool/actpool.go:488]: `if removeInvalid && ap.store != nil`

File: actpool/actpool.go [hunk 9]
```diff
@@ -412,6 +458,21 @@ func (ap *actPool) validate(ctx context.Context, selp *action.SealedEnvelope) er
...
+		hash, err := act.Hash()
+		if err != nil {
+			log.L().Debug("Skipping action due to hash error", zap.Error(err))
+			continue
+		}
+		log.L().Debug("Removed replaced action.", log.Hex("hash", hash[:]))
+		ap.allActions.Delete(hash)
+		intrinsicGas, _ := act.IntrinsicGas()
+		atomic.AddUint64(&ap.gasInPool, ^uint64(intrinsicGas-1))
+		ap.accountDesActs.delete(act)
```
Comment (dustinxie) [inline actpool/actpool.go:472]: for replaced acts, no need to delete from `ap.store`?

File: actpool/actionstore.go (hunk 3 above)
Comment (envestcc) [reply actpool/actionstore.go:73]: ok, maybe 1GB is enough

File: actpool/actpool.go (hunk 6 above)
Comment (envestcc) [reply actpool/actpool.go:461]: as discussed, will delete the `removeReplacedActs`

**Review by dustinxie — APPROVED** [review]

File: actpool/actpool.go [hunk 10]
```diff
@@ -153,6 +158,33 @@ func NewActPool(g genesis.Genesis, sf protocol.StateReader, cfg Config, opts ...
 	return ap, nil
 }
 
+func (ap *actPool) Start(ctx context.Context) error {
+	// open action store and load all actions
+	if ap.store != nil {
```
Comment (CoderZhi) [inline actpool/actpool.go:163]: early return

**Pull request merged**